Code/RDGeneral/RDValue-doublemagic.h

```diff
@@ -411,6 +410,22 @@ inline bool rdvalue_cast<bool>(RDValue v) {
   throw boost::bad_any_cast();
 }
 
+// The RDValue has no getKey implementation
+//  so implement our own
+struct KeyIntPair {
+  int key;
+  RDValue val;
+  KeyIntPair() : key(), val() {}
+  KeyIntPair(int k, const RDValue &v) : key(k), val(v) {}
+  KeyIntPair(const std::string &s, RDValue_cast_t v) :
+  key(tagmap.get(s)), val(v) {}
+  
+  inline void setKey(int k) { key = k; }
+  inline int  getKey() const { return key; }
+
+  static RDTags tagmap;
```

Comment:
It's totally reasonable for a user of RDKit to cache SMARTS patterns in a file-scope variable. I think a test case like this will raise our confidence. A test case like this will help catch such problems (I'm assuming SMARTS and SMILES trigger some properties, I'm not sure, maybe edit the SMILES or SMARTS to increase property usage): 

```
boost::shared_ptr<RDKit::ROMol> pat(RDKit::SmartsToMol("c1ccccc1"));
boost::shared_ptr<RDKit::RWMol> mol(RDKit::SmilesToMol("c1ccccc1O"));
int main() {
  MatchVectType res;
  TEST_ASSERT(SubstructMatch(*mol, *pat, res));
}
```